projects/igniteui-angular/src/lib/time-picker/time-picker.component.ts

```diff
@@ -2124,71 +1742,187 @@ export class IgxTimePickerComponent implements
-                setThroughUI: true
-            };
-            this.onValidationFailed.emit(args);
+            this.emitValidationFailedEvent(newValue, true);
         }
     }
 
-    private cursorOnHours(cursor: number, showHours: boolean): boolean {
-        return showHours && this._hoursPos.has(cursor);
+    private shouldCancelSelecting(): boolean {
+        const args: IBaseCancelableBrowserEventArgs = { owner: this, cancel: false };
+        this.selecting.emit(args);
+        return args.cancel;
     }
 
-    private cursorOnMinutes(cursor: number, showHours: boolean, showMinutes: boolean): boolean {
-        return showMinutes &&
-            (showHours && this._minutesPos.has(cursor)) ||
-            (!showHours && this._minutesPos.has(cursor));
+    private emitValueChangedEvent(oldValue: Date, newValue: Date) {
+        const args: IgxTimePickerValueChangedEventArgs = {
+            oldValue: oldValue,
+            newValue: newValue
+        };
+        this.valueChanged.emit(args);
     }
 
-    private cursorOnSeconds(cursor: number, showHours: boolean, showMinutes: boolean, showSeconds: boolean): boolean {
-        return showSeconds &&
-            (showHours && showMinutes && this._secondsPos.has(cursor)) ||
-            ((!showHours || !showMinutes) && this._secondsPos.has(cursor)) ||
-            (!showHours && !showMinutes && this._secondsPos.has(cursor));
+    private emitValidationFailedEvent(currentValue: Date, setThroughUI = false) {
+        const args: IgxTimePickerValidationFailedEventArgs = {
+            timePicker: this,
+            currentValue: currentValue,
+            setThroughUI: setThroughUI
+        };
+        this.validationFailed.emit(args);
     }
 
-    private cursorOnAmPm(cursor: number, showHours: boolean, showMinutes: boolean,
-        showSeconds: boolean, showAmPm: boolean): boolean {
-        return showAmPm &&
-            (showHours && showMinutes && showSeconds && this._amPmPos.has(cursor)) ||
-            ((!showHours || !showMinutes || !showSeconds) && this._amPmPos.has(cursor)) ||
-            (!showHours && (!showMinutes || !showSeconds) && this._amPmPos.has(cursor));
+    private updateValidityOnBlur() {
+        this._onTouchedCallback();
+        if (this.inputDirective && this._ngControl) {
+            if (!this._ngControl.valid) {
+                this.inputDirective.valid = IgxInputState.INVALID;
+            } else {
+                this.inputDirective.valid = IgxInputState.INITIAL;
+            }
+        }
     }
 
-    private _updateValidityOnBlur() {
-        this._onTouchedCallback();
-        const input = this._inputDirective || this._inputDirectiveUserTemplate;
-        if (this._ngControl && !this._ngControl.valid) {
-            input.valid = IgxInputState.INVALID;
-        } else {
-            input.valid = IgxInputState.INITIAL;
+    private valueInRange(value: Date, minValue?: Date, maxValue?: Date): boolean {
+        if (!value) {
```

Comment:
The check for non-existent value should be on the level above this method. Also, if there is no value, then it cannot be in a specified range. I.E. if there is no value, we have to make the picker invalid and this check here will return `true` if there is no value, implying that the picker is valid when it's empty, which is not the case if it's marked as `required`.